roslibrust/src/ros1/service_client.rs

```diff
@@ -0,0 +1,172 @@
+                return Err(err);
+            }
+            Err(_err) => {
+                return Err(RosLibRustError::Disconnected);
+            }
+        }
+    }
+}
+
+pub struct ServiceServerLink {
+    service_type: String,
+    call_sender: mpsc::UnboundedSender<CallServiceRequest>,
+    _actor_task: ChildTask<()>,
+}
+
+impl ServiceServerLink {
+    pub async fn new(
+        node_name: &Name,
+        service_name: &str,
+        service_type: &str,
+        service_uri: &str,
+        srv_definition: &str,
+        md5sum: &str,
+    ) -> RosLibRustResult<Self> {
+        let header = ConnectionHeader {
+            caller_id: node_name.to_string(),
+            latching: false,
+            msg_definition: srv_definition.to_owned(),
+            md5sum: md5sum.to_owned(),
+            topic: service_name.to_owned(),
+            topic_type: service_type.to_owned(),
+            tcp_nodelay: false,
+        };
+
+        let (call_tx, call_rx) = mpsc::unbounded_channel::<CallServiceRequest>();
+
+        let stream = establish_connection(&node_name, &service_name, &service_uri, header).await.map_err(|err| {
+            log::error!("Failed to establish connection to service URI {service_uri} for service {service_name}: {err}");
+            RosLibRustError::from(err)
+        })?;
+
+        let actor_context = Self::actor_context(stream, service_name.to_owned(), call_rx);
+
+        let handle = tokio::spawn(actor_context);
+
+        Ok(Self {
+            service_type: service_type.to_owned(),
+            call_sender: call_tx,
+            _actor_task: handle.into(),
+        })
+    }
+
+    pub fn get_sender(&self) -> mpsc::UnboundedSender<CallServiceRequest> {
+        self.call_sender.clone()
+    }
+
+    pub fn service_type(&self) -> &str {
+        &self.service_type
+    }
+
+    async fn actor_context(
+        mut stream: TcpStream,
+        service_name: String,
+        mut call_rx: UnboundedReceiver<CallServiceRequest>,
+    ) {
+        // Listen on a receiver for calls to forward to the service
+        loop {
+            match call_rx.recv().await {
+                Some(request) => {
+                    Self::handle_service_call(&mut stream, &service_name, request).await
+                }
+                None => {
+                    // Channel closed
+                    break;
+                }
+            }
+        }
+    }
+
+    async fn handle_service_call(
+        stream: &mut TcpStream,
+        service_name: &str,
+        (request, response_sender): CallServiceRequest,
+    ) {
+        match stream.write_all(&request).await {
+            Ok(()) => {
+                // Wait for the result
+                let mut read_buffer = vec![];
+                let read_result = match stream.read_buf(&mut read_buffer).await {
+                    // TODO: We may not get the full payload back in the call and need to check the length bytes
+                    Ok(_nbytes) => Ok(read_buffer),
```

Comment:
This is a very naive implementation and needs to be made robust for large payloads.